File: poc/MultiViewFeaturePoC/MultiViewFeaturePoC/Services/WindowManagerService.cs
using System;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading.Tasks;
using Windows.ApplicationModel.Core;
using Windows.UI.Core;
using Windows.UI.ViewManagement;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;

namespace MultiViewFeaturePoC.Services
{
    public delegate void ViewClosedHandler(ViewLifetimeControl viewControl, EventArgs e);

    public static class WindowManagerService
    {
        public static event ViewClosedHandler OnWindowClose;
        public static readonly Window MainWindow;

        public static readonly ApplicationView MainView;

        public static readonly ObservableCollection<ViewLifetimeControl> SecondaryViews;

        public static CoreWindow CurrentWindow => CoreWindow.GetForCurrentThread();

        public static CoreApplicationView CurrentView => CoreApplication.GetCurrentView();

        static WindowManagerService()
        {
            MainView = ApplicationView.GetForCurrentView();
            MainWindow = Window.Current;
            SecondaryViews = new ObservableCollection<ViewLifetimeControl>();
        }

        public static async Task<ViewLifetimeControl> TryShowAsStandaloneAsync(string windowTitle, Type pageType, object parameter = null, Action onCloseAction = null)
        {
            ViewLifetimeControl viewControl = SecondaryViews.FirstOrDefault(x => x.Title == windowTitle);
            if (viewControl == null)
            {
                viewControl = await CreateNewView(windowTitle, pageType, parameter, onCloseAction);
                viewControl.StartViewInUse();
                var viewShown = await ApplicationViewSwitcher.TryShowAsStandaloneAsync(viewControl.ViewId);
                viewControl.StopViewInUse();
            }
            else
            {
                await SwitchAsync(viewControl);
            }
            return viewControl;
        }

        public static async Task<ViewLifetimeControl> TryShowAsViewModeAsync(string windowTitle, Type pageType, object parameter = null, ApplicationViewMode applicationViewMode = ApplicationViewMode.CompactOverlay, Action onCloseAction = null)
        {
            ViewLifetimeControl viewControl = SecondaryViews.FirstOrDefault(x => x.Title == windowTitle);
            if (viewControl == null)
            {
                viewControl = await CreateNewView(windowTitle, pageType, parameter, onCloseAction);
                viewControl.StartViewInUse();
                var viewShown = await ApplicationViewSwitcher.TryShowAsViewModeAsync(viewControl.ViewId, applicationViewMode);
                viewControl.StopViewInUse();
            }
            else
            {
                await SwitchAsync(viewControl);
            }
            return viewControl;
        }

        public static async Task RunOnMainThreadAsync(CoreDispatcherPriority priority, DispatchedHandler agileCallback) => await MainWindow.Dispatcher.RunAsync(priority, agileCallback);

        public static async Task RunOnCurrentThreadAsync(CoreDispatcherPriority priority, DispatchedHandler agileCallback) => await CurrentWindow.Dispatcher.RunAsync(priority, agileCallback);

        public static bool IsWindowOpen(string windowTitle) => SecondaryViews.Any(v => v.Title == windowTitle);

        private static async Task<ViewLifetimeControl> CreateNewView(string windowTitle, Type pageType, object parameter = null, Action onCloseAction = null)
        {
            ViewLifetimeControl viewControl = null;
            var newView = CoreApplication.CreateNewView();
            await newView.Dispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
            {
                viewControl = ViewLifetimeControl.CreateForCurrentView();
                viewControl.Title = windowTitle;
                viewControl.OnCloseAction = onCloseAction;
                viewControl.Released += ViewControl_Released;
                viewControl.StartViewInUse();
                var frame = new Frame();
                frame.Navigate(pageType, viewControl);
                Window.Current.Content = frame;
                Window.Current.Activate();
                var applicationview = ApplicationView.GetForCurrentView();
                applicationview.Title = viewControl.Title;
            });
            SecondaryViews.Add(viewControl);
            return viewControl;
        }

        private static async Task SwitchAsync(ViewLifetimeControl viewControl)
        {
            await viewControl.Dispatcher.RunAsync(CoreDispatcherPriority.Normal, async () =>
            {
                await ApplicationViewSwitcher.SwitchAsync(viewControl.ViewId, MainView.Id, ApplicationViewSwitchingOptions.Default);
            });
        }

        private static void ViewControl_Released(object sender, EventArgs e)
        {
            var viewControl = sender as ViewLifetimeControl;
            viewControl.Released -= ViewControl_Released;
            SecondaryViews.Remove(viewControl);
            OnWindowClose?.Invoke(viewControl, EventArgs.Empty);
        }
    }
}

File: templates/Features/MultiViewVB/Services/WindowManagerService.vb
Imports Windows.ApplicationModel.Core
Imports Windows.UI.Core

Namespace Services

    Public Delegate Sub ViewClosedHandler(viewControl As ViewLifetimeControl, e As EventArgs)

    Public Class WindowManagerService

        ' TODO WTS: See more details about showing multiple views for an app
        ' https://docs.microsoft.com/windows/uwp/design/layout/show-multiple-views
        Private Shared _current As WindowManagerService

        Public Shared ReadOnly Property Current As WindowManagerService
            Get
                If _current Is Nothing Then
                    _current = New WindowManagerService()
                End If

                Return _current
            End Get
        End Property

        ' Contains all the opened secondary views.
        Public Property SecondaryViews As ObservableCollection(Of ViewLifetimeControl) = New ObservableCollection(Of ViewLifetimeControl)()

        Public Property MainViewId As Integer

        Public Property MainDispatcher As CoreDispatcher

        Public Sub Initialize()
            MainViewId = ApplicationView.GetForCurrentView().Id
            MainDispatcher = Window.Current.Dispatcher
        End Sub

        ' TODO WTS: Displays a view as a standalone
        ' You can use the resulting ViewLifeTileControl to interact with the new window.
        Public Async Function TryShowAsStandaloneAsync(windowTitle As String, pageType As Type) As Task(Of ViewLifetimeControl)
            Dim viewControl As ViewLifetimeControl = Await CreateViewLifetimeControlAsync(windowTitle, pageType)
            SecondaryViews.Add(viewControl)
            viewControl.StartViewInUse()
            Dim viewShown = Await ApplicationViewSwitcher.TryShowAsStandaloneAsync(viewControl.Id, ViewSizePreference.[Default], ApplicationView.GetForCurrentView().Id, ViewSizePreference.[Default])
            viewControl.StopViewInUse()
            Return viewControl
        End Function

        ' Displays a view as a standalone view in the desired view mode
        Public Async Function TryShowAsViewModeAsync(windowTitle As String, pageType As Type, Optional viewMode As ApplicationViewMode = ApplicationViewMode.[Default]) As Task(Of ViewLifetimeControl)
            Dim viewControl As ViewLifetimeControl = Await CreateViewLifetimeControlAsync(windowTitle, pageType)
            SecondaryViews.Add(viewControl)
            viewControl.StartViewInUse()
            Dim viewShown = Await ApplicationViewSwitcher.TryShowAsViewModeAsync(viewControl.Id, viewMode)
            viewControl.StopViewInUse()
            Return viewControl
        End Function

        Private Async Function CreateViewLifetimeControlAsync(windowTitle As String, pageType As Type) As Task(Of ViewLifetimeControl)
            Dim viewControl As ViewLifetimeControl = Nothing
            Await CoreApplication.CreateNewView().Dispatcher.RunAsync(CoreDispatcherPriority.Normal, Sub()
                viewControl = ViewLifetimeControl.CreateForCurrentView()
                viewControl.Title = windowTitle
                viewControl.StartViewInUse()
                Dim frame = New Frame()
                frame.Navigate(pageType, viewControl)
                Window.Current.Content = frame
                Window.Current.Activate()
                ApplicationView.GetForCurrentView().Title = viewControl.Title
            End Sub)
            Return viewControl
        End Function

        Public Function IsWindowOpen(windowTitle As String) As Boolean
            Return SecondaryViews.Any(Function(v) v.Title = windowTitle)
        End Function
    End Class
End Namespace
